linkerd/identity/src/imp/openssl.rs

```diff
@@ -0,0 +1,345 @@
+        let cert = crt.cert.clone();
+        if cert
+            .subject_alt_names()
+            .unwrap()
+            .iter()
+            .filter(|n| n.dnsname().unwrap().to_string() == crt.id.to_string())
+            .next()
+            .is_none()
+        {
+            return Err(InvalidCrt::SubjectAltName(crt.id.to_string()));
+        }
+
+        let mut chain = Stack::new().unwrap();
+        chain.push(cert.clone()).unwrap();
+        for chain_crt in crt.chain.clone() {
+            chain.push(chain_crt).unwrap();
+        }
+
+        let mut context = X509StoreContext::new().unwrap();
+        match context.init(&self.0, &cert, &chain, |c| match c.verify_cert() {
+            Ok(true) => Ok(Ok(true)),
+            Ok(false) => Ok(Err(InvalidCrt::Verify(c.error()))),
+            Err(err) => Err(err),
+        }) {
+            Ok(verify) => {
+                let server_config =
+                    ServerConfig::new(vec![], self.0.clone(), Some(crt.clone()), Some(key.clone()));
+                let client_config =
+                    ClientConfig::new(vec![], self.0.clone(), Some(crt.clone()), Some(key.clone()));
+
+                match verify {
+                    Ok(_) => Ok(CrtKey {
+                        id: crt.id.clone(),
+                        expiry: crt.expiry.clone(),
+                        client_config: Arc::new(client_config),
+                        server_config: Arc::new(server_config),
+                    }),
+                    Err(err) => Err(err),
+                }
+            }
+            Err(err) => Err(err.into()),
+        }
+    }
+
+    pub fn client_config(&self) -> Arc<ClientConfig> {
+        println!("Returning empty client config");
+        Arc::new(ClientConfig::new(vec![], self.0.clone(), None, None))
+    }
+}
+
+impl fmt::Debug for TrustAnchors {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "OpenSSL trust anchors")
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum InvalidCrt {
+    SubjectAltName(String),
+    Verify(X509VerifyResult),
+    General(Error),
+}
+
+impl From<Error> for InvalidCrt {
+    fn from(err: Error) -> Self {
+        InvalidCrt::General(err)
+    }
+}
+
+impl From<ErrorStack> for InvalidCrt {
+    fn from(err: ErrorStack) -> Self {
+        InvalidCrt::General(err.into())
+    }
+}
+
+impl fmt::Display for InvalidCrt {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            InvalidCrt::SubjectAltName(name) => write!(f, "Subject alt name incorrect {}", name),
+            InvalidCrt::Verify(err) => err.fmt(f),
+            InvalidCrt::General(err) => err.fmt(f),
+        }
+    }
+}
+
+impl error::Error for InvalidCrt {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        match self {
+            InvalidCrt::Verify(err) => err.source(),
+            InvalidCrt::General(err) => err.source(),
+            _ => None,
+        }
+    }
+}
```

Comment:
nit: again, I think a lot of this boilerplate could be generated using `thiserror`, like so
```suggestion
#[derive(Clone, Debug. Error)]
pub enum InvalidCrt {
    #[error("subject alt name incorrect ({0})"]
    SubjectAltName(String),
    #[error("{0}")]
    Verify(#[source] X509VerifyResult),
    #[error(transparent)]
    General(#[from] Error),
}

impl From<ErrorStack> for InvalidCrt {
    fn from(err: ErrorStack) -> Self {
        InvalidCrt::General(err.into())
    }
}
```